File: src/main/java/cadenza/Detector.java
package cadenza;

import com.oracle.truffle.api.TruffleFile;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.regex.Pattern;

public class Detector implements TruffleFile.FileTypeDetector {
  private static final Pattern SHEBANG_REGEXP = Pattern.compile("^#! ?/usr/bin/(env +ccc|ccc).*");
  @Override public String findMimeType(TruffleFile file) {
    String name = file.getName();
    if (name == null) return null;
    if (name.endsWith(Language.EXTENSION)) return Language.MIME_TYPE;
    try (BufferedReader fileContent = file.newBufferedReader(StandardCharsets.UTF_8)) {
      final String firstLine = fileContent.readLine();
      if (firstLine != null && SHEBANG_REGEXP.matcher(firstLine).matches())
        return Language.MIME_TYPE;
    } catch (IOException | SecurityException e) {
      // ok
    }
    return null;
  }

  @Override public Charset findEncoding(TruffleFile file) {
    return StandardCharsets.UTF_8;
  }
}

File: language/src/main/kotlin/cadenza/Detector.kt
package cadenza

import com.oracle.truffle.api.TruffleFile

import java.io.BufferedReader
import java.io.IOException
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.util.regex.Pattern

class Detector : TruffleFile.FileTypeDetector {
  override fun findMimeType(file: TruffleFile): String? {
    val name = file.name ?: return null
    if (name.endsWith(Language.EXTENSION)) return Language.MIME_TYPE
    try {
      file.newBufferedReader(StandardCharsets.UTF_8).use { fileContent ->
        val firstLine = fileContent.readLine()
        if (firstLine != null && SHEBANG_REGEXP.matcher(firstLine).matches())
          return Language.MIME_TYPE
      }
    } catch (e: IOException) {
      // ok
    } catch (e: SecurityException) {
    }

    return null
  }

  override fun findEncoding(_file: TruffleFile): Charset {
    return StandardCharsets.UTF_8
  }

  companion object {
    private val SHEBANG_REGEXP = Pattern.compile("^#! ?/usr/bin/(env +ccc|ccc).*")
  }
}
